# import libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import sklearn as sk

/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


# import dataset

In [2]:
path='/kaggle/input/-cerial-prices-changes-within-last-30-years/rice_wheat_corn_prices.csv'
df=pd.read_csv(path)
df.head()

Year Month  Price_wheat_ton  Price_rice_ton  Price_corn_ton  \
0  1992   Feb           170.12          278.25          113.62   
1  1992   Mar           161.44          277.20          117.00   
2  1992   Apr           153.07          278.00          108.52   
3  1992   May           139.72          274.00          109.64   
4  1992   Jun           140.36          268.80          110.90   

   Inflation_rate  Price_wheat_ton_infl  Price_rice_ton_infl  \
0           89.59                322.53               527.53   
1           89.59                306.07               525.54   
2           89.59                290.21               527.06   
3           89.59                264.90               519.48   
4           89.59                266.11               509.62   

   Price_corn_ton_infl  
0               215.41  
1               221.82  
2               205.74  
3               207.87  
4               210.26

# data cleaning and preparing

# check of white spaces in columns name

In [3]:
df.columns

Index(['Year', 'Month', 'Price_wheat_ton', 'Price_rice_ton', 'Price_corn_ton',
       'Inflation_rate', 'Price_wheat_ton_infl', 'Price_rice_ton_infl',
       'Price_corn_ton_infl'],
      dtype='object')

In [4]:
df.columns=df.columns.str.strip()
df.columns

Index(['Year', 'Month', 'Price_wheat_ton', 'Price_rice_ton', 'Price_corn_ton',
       'Inflation_rate', 'Price_wheat_ton_infl', 'Price_rice_ton_infl',
       'Price_corn_ton_infl'],
      dtype='object')

# check of data types

In [5]:
df.dtypes

Year                      int64
Month                    object
Price_wheat_ton         float64
Price_rice_ton          float64
Price_corn_ton          float64
Inflation_rate          float64
Price_wheat_ton_infl    float64
Price_rice_ton_infl     float64
Price_corn_ton_infl     float64
dtype: object

**change the 'Month' column to int later in analyzing process**

# check null values

In [6]:
df.isnull().sum()

Year                    0
Month                   0
Price_wheat_ton         1
Price_rice_ton          0
Price_corn_ton          0
Inflation_rate          1
Price_wheat_ton_infl    1
Price_rice_ton_infl     1
Price_corn_ton_infl     1
dtype: int64

In [7]:
# try to find missing data by another valid one

In [8]:
df[df['Price_wheat_ton'].isnull()]

Year Month  Price_wheat_ton  Price_rice_ton  Price_corn_ton  \
359  2022   Jan              NaN           427.0          276.62   

     Inflation_rate  Price_wheat_ton_infl  Price_rice_ton_infl  \
359             NaN                   NaN                  NaN   

     Price_corn_ton_infl  
359                  NaN

In [9]:
df[(df['Year']==2022) & (df['Month']=='Jan')]

Year Month  Price_wheat_ton  Price_rice_ton  Price_corn_ton  \
359  2022   Jan              NaN           427.0          276.62   

     Inflation_rate  Price_wheat_ton_infl  Price_rice_ton_infl  \
359             NaN                   NaN                  NaN   

     Price_corn_ton_infl  
359                  NaN

In [10]:
df[df['Inflation_rate'].isnull()]

Year Month  Price_wheat_ton  Price_rice_ton  Price_corn_ton  \
359  2022   Jan              NaN           427.0          276.62   

     Inflation_rate  Price_wheat_ton_infl  Price_rice_ton_infl  \
359             NaN                   NaN                  NaN   

     Price_corn_ton_infl  
359                  NaN

In [11]:
# drop null values

In [12]:
df.dropna(axis=0,inplace=True)

In [13]:
df.isnull().sum()

Year                    0
Month                   0
Price_wheat_ton         0
Price_rice_ton          0
Price_corn_ton          0
Inflation_rate          0
Price_wheat_ton_infl    0
Price_rice_ton_infl     0
Price_corn_ton_infl     0
dtype: int64

# check duplicated values

In [14]:
df.duplicated().sum()

0

# check outliers

In [15]:
df.describe()

Year  Price_wheat_ton  Price_rice_ton  Price_corn_ton  \
count   359.000000       359.000000      359.000000      359.000000   
mean   2006.540390       185.302869      363.930418      155.165376   
std       8.645592        64.985279      131.508817       62.370286   
min    1992.000000        85.300000      163.750000       75.270000   
25%    1999.000000       137.315000      261.250000      104.155000   
50%    2007.000000       175.270000      343.750000      149.340000   
75%    2014.000000       220.260000      444.500000      176.010000   
max    2021.000000       419.610000      907.000000      333.050000   

       Inflation_rate  Price_wheat_ton_infl  Price_rice_ton_infl  \
count      359.000000            359.000000           359.000000   
mean        36.316685            241.726769           474.038384   
std         26.605378             65.234300           133.776144   
min         -1.290000            136.220000           246.020000   
25%         12.370000            193.010000           397.500000   
50%         28.300000            228.310000           455.540000   
75%         59.700000            275.835000           540.180000   
max         89.590000            518.470000          1120.690000   

       Price_corn_ton_infl  
count           359.000000  
mean            201.211086  
std              61.727752  
min             116.280000  
25%             159.015000  
50%             186.020000  
75%             220.940000  
max             385.910000

In [16]:
df.columns

Index(['Year', 'Month', 'Price_wheat_ton', 'Price_rice_ton', 'Price_corn_ton',
       'Inflation_rate', 'Price_wheat_ton_infl', 'Price_rice_ton_infl',
       'Price_corn_ton_infl'],
      dtype='object')

In [17]:
numerical_df=df.drop(columns='Month',axis=1)
numerical_df.head()

Year  Price_wheat_ton  Price_rice_ton  Price_corn_ton  Inflation_rate  \
0  1992           170.12          278.25          113.62           89.59   
1  1992           161.44          277.20          117.00           89.59   
2  1992           153.07          278.00          108.52           89.59   
3  1992           139.72          274.00          109.64           89.59   
4  1992           140.36          268.80          110.90           89.59   

   Price_wheat_ton_infl  Price_rice_ton_infl  Price_corn_ton_infl  
0                322.53               527.53               215.41  
1                306.07               525.54               221.82  
2                290.21               527.06               205.74  
3                264.90               519.48               207.87  
4                266.11               509.62               210.26

In [18]:
fig=make_subplots(rows=1,cols=8)

n=1
for column in numerical_df.columns :
    fig.add_trace(go.Box(y=numerical_df[column],name=column,boxpoints='suspectedoutliers'),row=1,col=n)
    n=n+1
    
fig.show()

In [19]:
df[df['Price_rice_ton_infl'] > 1120]

Year Month  Price_wheat_ton  Price_rice_ton  Price_corn_ton  \
194  2008   Apr           323.43           907.0          246.44   

     Inflation_rate  Price_wheat_ton_infl  Price_rice_ton_infl  \
194           23.56                399.63              1120.69   

     Price_corn_ton_infl  
194                304.5

In [20]:
df.shape

(359, 9)

In [21]:
# drop 'Price_rice_ton_infl' > 1120

df=df[df['Price_rice_ton_infl'] < 1120]
df.shape

(358, 9)

In [22]:
numerical_df=df.drop(columns='Month',axis=1)
numerical_df.head()

Year  Price_wheat_ton  Price_rice_ton  Price_corn_ton  Inflation_rate  \
0  1992           170.12          278.25          113.62           89.59   
1  1992           161.44          277.20          117.00           89.59   
2  1992           153.07          278.00          108.52           89.59   
3  1992           139.72          274.00          109.64           89.59   
4  1992           140.36          268.80          110.90           89.59   

   Price_wheat_ton_infl  Price_rice_ton_infl  Price_corn_ton_infl  
0                322.53               527.53               215.41  
1                306.07               525.54               221.82  
2                290.21               527.06               205.74  
3                264.90               519.48               207.87  
4                266.11               509.62               210.26

In [23]:
fig=make_subplots(rows=1,cols=8)

n=1
for column in numerical_df.columns :
    fig.add_trace(go.Box(y=numerical_df[column],name=column,boxpoints='suspectedoutliers'),row=1,col=n)
    n=n+1
    
fig.show()

In [24]:
# drop record 'Price_rice_ton_infl' = 1114.26

df=df[df['Price_rice_ton_infl'] <= 935.35]
df.shape

(357, 9)

In [25]:
numerical_df=df.drop(columns='Month',axis=1)
numerical_df.head()

Year  Price_wheat_ton  Price_rice_ton  Price_corn_ton  Inflation_rate  \
0  1992           170.12          278.25          113.62           89.59   
1  1992           161.44          277.20          117.00           89.59   
2  1992           153.07          278.00          108.52           89.59   
3  1992           139.72          274.00          109.64           89.59   
4  1992           140.36          268.80          110.90           89.59   

   Price_wheat_ton_infl  Price_rice_ton_infl  Price_corn_ton_infl  
0                322.53               527.53               215.41  
1                306.07               525.54               221.82  
2                290.21               527.06               205.74  
3                264.90               519.48               207.87  
4                266.11               509.62               210.26

In [26]:
fig=make_subplots(rows=1,cols=8)

n=1
for column in numerical_df.columns :
    fig.add_trace(go.Box(y=numerical_df[column],name=column,boxpoints='suspectedoutliers'),row=1,col=n)
    n=n+1
    
fig.show()

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 357 entries, 0 to 358
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Year                  357 non-null    int64  
 1   Month                 357 non-null    object 
 2   Price_wheat_ton       357 non-null    float64
 3   Price_rice_ton        357 non-null    float64
 4   Price_corn_ton        357 non-null    float64
 5   Inflation_rate        357 non-null    float64
 6   Price_wheat_ton_infl  357 non-null    float64
 7   Price_rice_ton_infl   357 non-null    float64
 8   Price_corn_ton_infl   357 non-null    float64
dtypes: float64(7), int64(1), object(1)
memory usage: 27.9+ KB


# data analysis

In [28]:
df.head()

Year Month  Price_wheat_ton  Price_rice_ton  Price_corn_ton  \
0  1992   Feb           170.12          278.25          113.62   
1  1992   Mar           161.44          277.20          117.00   
2  1992   Apr           153.07          278.00          108.52   
3  1992   May           139.72          274.00          109.64   
4  1992   Jun           140.36          268.80          110.90   

   Inflation_rate  Price_wheat_ton_infl  Price_rice_ton_infl  \
0           89.59                322.53               527.53   
1           89.59                306.07               525.54   
2           89.59                290.21               527.06   
3           89.59                264.90               519.48   
4           89.59                266.11               509.62   

   Price_corn_ton_infl  
0               215.41  
1               221.82  
2               205.74  
3               207.87  
4               210.26

In [29]:
df['Month'].unique()

array(['Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct',
       'Nov', 'Dec', 'Jan'], dtype=object)

In [30]:
df['Month_num']=df['Month']

In [31]:
df.head()

Year Month  Price_wheat_ton  Price_rice_ton  Price_corn_ton  \
0  1992   Feb           170.12          278.25          113.62   
1  1992   Mar           161.44          277.20          117.00   
2  1992   Apr           153.07          278.00          108.52   
3  1992   May           139.72          274.00          109.64   
4  1992   Jun           140.36          268.80          110.90   

   Inflation_rate  Price_wheat_ton_infl  Price_rice_ton_infl  \
0           89.59                322.53               527.53   
1           89.59                306.07               525.54   
2           89.59                290.21               527.06   
3           89.59                264.90               519.48   
4           89.59                266.11               509.62   

   Price_corn_ton_infl Month_num  
0               215.41       Feb  
1               221.82       Mar  
2               205.74       Apr  
3               207.87       May  
4               210.26       Jun

In [32]:
for x in df.index :
    if df.loc[x,'Month_num']=='Feb' :
        df.loc[x,'Month_num']=2
    elif df.loc[x,'Month_num']=='Mar' :
        df.loc[x,'Month_num']=3
    elif df.loc[x,'Month_num']=='Apr' :
        df.loc[x,'Month_num']=4
    elif df.loc[x,'Month_num']=='May' :
        df.loc[x,'Month_num']=5
    elif df.loc[x,'Month_num']=='Jun' :
        df.loc[x,'Month_num']=6
    elif df.loc[x,'Month_num']=='Jul' :
        df.loc[x,'Month_num']=7
    elif df.loc[x,'Month_num']=='Aug' :
        df.loc[x,'Month_num']=8
    elif df.loc[x,'Month_num']=='Sep' :
        df.loc[x,'Month_num']=9
    elif df.loc[x,'Month_num']=='Oct' :
        df.loc[x,'Month_num']=10
    elif df.loc[x,'Month_num']=='Nov' :
        df.loc[x,'Month_num']=11
    elif df.loc[x,'Month_num']=='Dec' :
        df.loc[x,'Month_num']=12
    elif df.loc[x,'Month_num']=='Jan' :
        df.loc[x,'Month_num']=1

In [33]:
df.head()

Year Month  Price_wheat_ton  Price_rice_ton  Price_corn_ton  \
0  1992   Feb           170.12          278.25          113.62   
1  1992   Mar           161.44          277.20          117.00   
2  1992   Apr           153.07          278.00          108.52   
3  1992   May           139.72          274.00          109.64   
4  1992   Jun           140.36          268.80          110.90   

   Inflation_rate  Price_wheat_ton_infl  Price_rice_ton_infl  \
0           89.59                322.53               527.53   
1           89.59                306.07               525.54   
2           89.59                290.21               527.06   
3           89.59                264.90               519.48   
4           89.59                266.11               509.62   

   Price_corn_ton_infl Month_num  
0               215.41         2  
1               221.82         3  
2               205.74         4  
3               207.87         5  
4               210.26         6

# How much AVG Inflation_rate regarding year?

In [34]:
df[['Year','Inflation_rate']].groupby('Year').agg('mean')

Inflation_rate
Year                
1992           89.59
1993           84.15
1994           79.47
1995           74.58
1996           69.60
1997           65.73
1998           63.20
1999           59.70
2000           54.49
2001           50.24
2002           47.89
2003           44.61
2004           40.84
2005           36.22
2006           31.96
2007           28.30
2008           23.56
2009           24.00
2010           22.00
2011           18.27
2012           15.87
2013           14.19
2014           12.37
2015           12.24
2016           10.84
2017            8.53
2018            5.94
2019            4.06
2020            2.79
2021           -1.29

In [35]:
px.line(df[['Year','Inflation_rate']].groupby('Year').agg('mean'),markers=True)

In [36]:
df['Year']=df['Year'].astype(str)
df['Year']

0      1992
1      1992
2      1992
3      1992
4      1992
       ... 
354    2021
355    2021
356    2021
357    2021
358    2021
Name: Year, Length: 357, dtype: object

In [37]:
df['Date']=df['Year']+'-'+df['Month']
df['Date']

0      1992-Feb
1      1992-Mar
2      1992-Apr
3      1992-May
4      1992-Jun
         ...   
354    2021-Aug
355    2021-Sep
356    2021-Oct
357    2021-Nov
358    2021-Dec
Name: Date, Length: 357, dtype: object

In [38]:
df['Date']=pd.to_datetime(df['Date'])
df['Date']

0     1992-02-01
1     1992-03-01
2     1992-04-01
3     1992-05-01
4     1992-06-01
         ...    
354   2021-08-01
355   2021-09-01
356   2021-10-01
357   2021-11-01
358   2021-12-01
Name: Date, Length: 357, dtype: datetime64[ns]

# How much Inflation_rate regarding Date?

In [39]:
df[['Date','Inflation_rate']].groupby('Date').agg('mean')

Inflation_rate
Date                      
1992-02-01           89.59
1992-03-01           89.59
1992-04-01           89.59
1992-05-01           89.59
1992-06-01           89.59
...                    ...
2021-08-01           -1.29
2021-09-01           -1.29
2021-10-01           -1.29
2021-11-01           -1.29
2021-12-01           -1.29

[357 rows x 1 columns]

In [40]:
px.line(df[['Date','Inflation_rate']].groupby('Date').agg('mean'))

**'Inflation rate' was always decreased except for dates between 2008 to 2009 it raised from 23.56 to 24 and in 2010 dropped again to 22**

# At which month of the year does the drop in the Inflation rate happened usually?

In [41]:
df['Year']=df['Year'].astype(int)
df['Year']

0      1992
1      1992
2      1992
3      1992
4      1992
       ... 
354    2021
355    2021
356    2021
357    2021
358    2021
Name: Year, Length: 357, dtype: int64

In [42]:
desc_df=df.pivot_table(index='Year',columns='Month_num',values='Inflation_rate',aggfunc='mean')
desc_df

Month_num     1      2      3      4      5      6      7      8      9   \
Year                                                                       
1992         NaN  89.59  89.59  89.59  89.59  89.59  89.59  89.59  89.59   
1993       84.15  84.15  84.15  84.15  84.15  84.15  84.15  84.15  84.15   
1994       79.47  79.47  79.47  79.47  79.47  79.47  79.47  79.47  79.47   
1995       74.58  74.58  74.58  74.58  74.58  74.58  74.58  74.58  74.58   
1996       69.60  69.60  69.60  69.60  69.60  69.60  69.60  69.60  69.60   
1997       65.73  65.73  65.73  65.73  65.73  65.73  65.73  65.73  65.73   
1998       63.20  63.20  63.20  63.20  63.20  63.20  63.20  63.20  63.20   
1999       59.70  59.70  59.70  59.70  59.70  59.70  59.70  59.70  59.70   
2000       54.49  54.49  54.49  54.49  54.49  54.49  54.49  54.49  54.49   
2001       50.24  50.24  50.24  50.24  50.24  50.24  50.24  50.24  50.24   
2002       47.89  47.89  47.89  47.89  47.89  47.89  47.89  47.89  47.89   
2003       44.61  44.61  44.61  44.61  44.61  44.61  44.61  44.61  44.61   
2004       40.84  40.84  40.84  40.84  40.84  40.84  40.84  40.84  40.84   
2005       36.22  36.22  36.22  36.22  36.22  36.22  36.22  36.22  36.22   
2006       31.96  31.96  31.96  31.96  31.96  31.96  31.96  31.96  31.96   
2007       28.30  28.30  28.30  28.30  28.30  28.30  28.30  28.30  28.30   
2008       23.56  23.56  23.56    NaN    NaN  23.56  23.56  23.56  23.56   
2009       24.00  24.00  24.00  24.00  24.00  24.00  24.00  24.00  24.00   
2010       22.00  22.00  22.00  22.00  22.00  22.00  22.00  22.00  22.00   
2011       18.27  18.27  18.27  18.27  18.27  18.27  18.27  18.27  18.27   
2012       15.87  15.87  15.87  15.87  15.87  15.87  15.87  15.87  15.87   
2013       14.19  14.19  14.19  14.19  14.19  14.19  14.19  14.19  14.19   
2014       12.37  12.37  12.37  12.37  12.37  12.37  12.37  12.37  12.37   
2015       12.24  12.24  12.24  12.24  12.24  12.24  12.24  12.24  12.24   
2016       10.84  10.84  10.84  10.84  10.84  10.84  10.84  10.84  10.84   
2017        8.53   8.53   8.53   8.53   8.53   8.53   8.53   8.53   8.53   
2018        5.94   5.94   5.94   5.94   5.94   5.94   5.94   5.94   5.94   
2019        4.06   4.06   4.06   4.06   4.06   4.06   4.06   4.06   4.06   
2020        2.79   2.79   2.79   2.79   2.79   2.79   2.79   2.79   2.79   
2021       -1.29  -1.29  -1.29  -1.29  -1.29  -1.29  -1.29  -1.29  -1.29   

Month_num     10     11     12  
Year                            
1992       89.59  89.59  89.59  
1993       84.15  84.15  84.15  
1994       79.47  79.47  79.47  
1995       74.58  74.58  74.58  
1996       69.60  69.60  69.60  
1997       65.73  65.73  65.73  
1998       63.20  63.20  63.20  
1999       59.70  59.70  59.70  
2000       54.49  54.49  54.49  
2001       50.24  50.24  50.24  
2002       47.89  47.89  47.89  
2003       44.61  44.61  44.61  
2004       40.84  40.84  40.84  
2005       36.22  36.22  36.22  
2006       31.96  31.96  31.96  
2007       28.30  28.30  28.30  
2008       23.56  23.56  23.56  
2009       24.00  24.00  24.00  
2010       22.00  22.00  22.00  
2011       18.27  18.27  18.27  
2012       15.87  15.87  15.87  
2013       14.19  14.19  14.19  
2014       12.37  12.37  12.37  
2015       12.24  12.24  12.24  
2016       10.84  10.84  10.84  
2017        8.53   8.53   8.53  
2018        5.94   5.94   5.94  
2019        4.06   4.06   4.06  
2020        2.79   2.79   2.79  
2021       -1.29  -1.29  -1.29

In [43]:
px.line(desc_df.loc[1992,:],markers=True)

In [44]:
px.line(desc_df.loc[1993,:],markers=True)

**the drop usually happened in Jan of each year**

# Recommendation check records

**I think this dataset is not real**